solidity/contracts/middleware/liquidity-layer/adapters/PortalAdapter.sol

```diff
@@ -0,0 +1,216 @@
+     * @param hyperlaneDomain The Hyperlane domain.
+     * @param womrholeDomain The Wormhole domain.
+     */
+    event DomainAdded(uint32 indexed hyperlaneDomain, uint32 womrholeDomain);
+
+    modifier onlyLiquidityLayerRouter() {
+        require(msg.sender == liquidityLayerRouter, "!liquidityLayerRouter");
+        _;
+    }
+
+    /**
+     * @param _localDomain The local hyperlane domain
+     * @param _owner The new owner.
+     * @param _portalTokenBridge The Portal TokenBridge contract.
+     * @param _liquidityLayerRouter The LiquidityLayerRouter contract.
+     */
+    function initialize(
+        uint32 _localDomain,
+        address _owner,
+        address _portalTokenBridge,
+        address _liquidityLayerRouter
+    ) public initializer {
+        // Transfer ownership of the contract to deployer
+        _transferOwnership(_owner);
+
+        // Set the addresses for the ACM and IGP to address(0) - they aren't used.
+        _setAbacusConnectionManager(address(0));
+        _setInterchainGasPaymaster(address(0));
+
+        localDomain = _localDomain;
+        portalTokenBridge = IPortalTokenBridge(_portalTokenBridge);
+        liquidityLayerRouter = _liquidityLayerRouter;
+    }
+
+    function sendTokens(
+        uint32 _destinationDomain,
+        bytes32, // _recipientAddress, unused
+        address _token,
+        uint256 _amount
+    ) external onlyLiquidityLayerRouter returns (bytes memory) {
+        nonce = nonce + 1;
+        uint16 wormholeDomain = hyperlaneDomainToWormholeDomain[
+            _destinationDomain
+        ];
+
+        bytes32 _remoteRouter = routers[_destinationDomain];
+        require(
+            _remoteRouter != bytes32(0),
+            "Portal TokenBridgeAdapter: No router for domain"
+        );
+
+        // Approve the token to Portal. We assume that the LiquidityLayerRouter
+        // has already transferred the token to this contract.
+        require(
+            IERC20(_token).approve(address(portalTokenBridge), _amount),
+            "!approval"
+        );
+
+        bytes memory payload = adapterData(localDomain, nonce);
+
+        uint64 portalSequence = portalTokenBridge.transferTokensWithPayload(
+            _token,
+            _amount,
+            wormholeDomain,
+            _remoteRouter,
+            0,
+            payload
+        );
+
+        emit BridgedToken(nonce, portalSequence, _destinationDomain);
+        return payload;
+    }
+
+    // Returns the token and amount sent
+    function receiveTokens(
+        uint32 _originDomain, // Hyperlane domain
+        address _recipient,
+        uint256 _amount,
+        bytes calldata _adapterData // The adapter data from the message
+    ) external onlyLiquidityLayerRouter returns (address, uint256) {
+        // Get the origin information from the adapterData
+        (uint32 _originDomainInPayload, uint224 _nonce) = abi.decode(
```

Comment:
Yeah I think I originally had that in there because its also the payload on the wormhole message and necessary to mark the portal transfer as complete, but now we just parse those from the whole portal transfer message

Reply:
Actually nvm, it doesn't look like the portal transfer message makes the origin domain accessible (which is surprising).